File: encoding/go/message.go
package message

import (
	"strings"
)

const (
	msgsep = "\r"
	msgeq  = ":"
)

// Encode formats a map and returns a string
func Encode(table map[string]string) string {
	var b strings.Builder

	for k, v := range table {
		b.WriteString(k + msgeq + v + msgsep)
	}
	return b.String()
}

// Decode parses a string expression and returns a map
func Decode(msg string) map[string]string {
	table := make(map[string]string)
	lines := strings.Split(msg, msgsep)
	for _, line := range lines {
		splits := strings.SplitN(line, msgeq, 2)
		if len(splits) < 2 {
			continue
		}
		k, v := splits[0], splits[1]
		table[k] = v
	}
	return table
}

File: encoding/message.cpp
#include "message.h"
#include <sstream>

static const char MSGSEP = '\r';
static const char MSGEND = '\n';
static const char MSGEQ = ':';

// encode
std::string encode(const std::unordered_map<std::string, std::string> &table) {
  std::stringstream ss;
  for (const auto &it : table) ss << it.first << MSGEQ << it.second << MSGSEP;
  ss << MSGEND;
  return ss.str();
}

// decode
std::unordered_map<std::string, std::string> decode(const std::string &msg) {
  std::unordered_map<std::string, std::string> table;

  std::string key, val;
  std::istringstream iss(msg);

  while (std::getline(std::getline(iss, key, MSGEQ), val, MSGSEP))
    table[key] = val;

  return table;
}